pypeclub/pype
# Black Magic DaVinci Resolve Host integration

**Overview**
DaVinci Resolve is great tool for editing, color-grading, media management and encoding. 
Adding Resolve into our portfolio of integrated hosts should not be a big issue as most of the functionality has  been written already for Premiere and Hiero. Last couple of years, some useful open source code using Resolve’s python api has been developed which could inspire our development.

**Goals**
- Be able to batch rename clips
- Be able to tag clips with families for further subset definitions in publishing
- Be able to batch export and publish conformed clips into pipeline
- Be able to batch load (integrate) from db subsets to timeline as version managed containers

**Problems**
- app GUI expand-ability limits
- python 3.6 has to be installed on particular place in system
- python 3.7 or higher not supported

**Milestones**
- Project context wrapper [ui]
- Pype menu [menu]
- Tools for batch renaming of clips inspired by Hiero [ui]
- Versioning of clips [timeline>menu>submenu]
- Publishing with Pype into Avalon DB and Ftrack
- Integrating / Loading subsets back to timeline as containerized clips

**Github projects**
https://github.com/Puppetworks-Animation-Studio/deadline-davinci-resolve
https://github.com/veryqiang/footbrake
https://github.com/IgorRidanovic/DaVinciResolve-ExportProjects
https://github.com/Benerman/DaVinci-Resolve-Scripting
https://github.com/pressreset/resutil

**Testing this PR**
A. _Installing Resolve's own python 3.6 interpreter._
DaVinci Resolve is using hard-coded method for searching python path and ignoring what ever is in PATH. Windows is searching for "{LOCALAPPDATA}/Programs/Python/Python36" on mac it is "~/Library/Python/3.6/bin" and linux "/opt/Python/3.6/bin". All of these paths are chosen automatically during Python 3.6 installer. We are using `Python 3.6.2`

B. _Installing PySide2 into python 3.6 for correct gui work:_
1. As Pype is using gui inside Resolve, we need to install own PySide2 into the python 3.6 as it is explained above. Follow this step-by-step:
2. open terminal > windows menu > type `cmd`
3. paste this to terminal > %LOCALAPPDATA%\Programs\Python\Python36\python.exe -m pip install PySide2
4. after installation finished then all is done

C. _Set Resolve's Fusion settings for Python 3.6 interpereter:_
As it is shown in bellow picture you have to go to Fusion Tab and then in Fusion menu find Fusion Settings. Go to Fusion/Script and find `Default Python Version` and swith to Python 3.6

D. _Set correct paths in pype-config_
- set env var **RESOLVE_UTILITY_SCRIPTS_SOURCE_DIR** in `environments\resolve.json` to a real empty directory or just keep the list empty.

E. _In Ftrack on project run Update Custom Attributes_
This will add Resolve launcher into available apps for use on a project
![image](https://user-images.githubusercontent.com/40640033/83256237-5de8a600-a1b2-11ea-89bb-e46dcd1f53e7.png)

F. _In Ftrack on project info panel in Applications attribute select BM DaVinci Resolve 16_
![image](https://user-images.githubusercontent.com/40640033/83256148-31348e80-a1b2-11ea-981d-309668c5fe7c.png)

G. _In Ftrack on project Actions chose SyncToAvalon action_
![image](https://user-images.githubusercontent.com/40640033/83256286-7b1d7480-a1b2-11ea-9d28-88fa5ed83b37.png)

H. _In Ftrack on project create directory `editorial` and task `conforming`, save the Ftrack to db and repeate G. again_
![image](https://user-images.githubusercontent.com/40640033/83256523-da7b8480-a1b2-11ea-8b01-32a555130e99.png)

I. _Start BM DaVinci Resolve 16 from actions on the task `conforming`_
![image](https://user-images.githubusercontent.com/40640033/83256630-0ac32300-a1b3-11ea-87d2-f7444ee0fa1e.png)

If you would like to know more about ongoing development just follow this link: https://miro.com/app/board/o9J_kt15XIY=/ 
![image](https://user-images.githubusercontent.com/40640033/83248664-ba919400-a1a5-11ea-8068-1d878a4a7d0a.png)

resolving #96 
dependent PR: https://github.com/pypeclub/pype-config/pull/33
